File: compiler/src/feedback/items/NameFeedback.swift
import OrderedCollections

func registerNameFeedback(into context: FeedbackContext) {
    context.register(
        id: "unresolved",
        query: { (db, node, body: (Node, String, (node: Node, name: String)?) -> Void) in
            Queries.unresolved(db, node) { name, suggestion in body(node, name, suggestion) }
        },
        rank: { _, _, _ in .names },
        location: { node, _, suggestion in
            let secondary = suggestion.map { OrderedSet([$0.node]) } ?? []
            return (primary: node, secondary: secondary)
        },
    ) { writer, _, name, suggestion in
        writer.write("Can't find ")
        writer.write(code: name)
        writer.write(".")
        writer.writeBreak()

        if let suggestion {
            writer.write("Did you mean ")
            writer.writeLink(suggestion.name, suggestion.node)
            writer.write("?")
        } else {
            writer.write("Double-check your spelling.")
        }
    }

    context.register(
        id: "ambiguous",
        query: Queries.ambiguous,
        rank: { _, _, _ in .names },
        location: { node, _, _ in (primary: node, secondary: []) },
    ) { writer, _, name, definitions in
        writer.write(code: name)
        writer.write(" could refer to ")
        writer.writeList(separator: "or") { list in
            for definition in definitions { list.add { writer.write(definition) } }
        }
        writer.write(".")
        writer.writeBreak()
        writer.write("Rename the extra definitions.")
    }
}

File: compiler/feedback/src/items/name_feedback.rs
use crate::{FeedbackCtx, FeedbackLocation, FeedbackRank};
use std::collections::BTreeSet;
use wipple_queries::{ambiguous, unresolved};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("unresolved")
        .query(unresolved)
        .rank(|_| FeedbackRank::Names)
        .location(|node, (_, suggestion)| FeedbackLocation {
            primary: node,
            secondary: suggestion
                .as_ref()
                .map(|(node, _)| BTreeSet::from([*node]))
                .unwrap_or_default(),
        })
        .display(|_db, writer, _, (name, suggestion)| {
            writer.string("Can't find ");
            writer.code(name.to_string());
            writer.string(".");
            writer.line_break();

            if let Some((node, name)) = suggestion {
                writer.string("Did you mean ");
                writer.link(name.to_string(), *node);
                writer.string("?");
            } else {
                writer.string("Double-check your spelling.");
            }
        })
        .register();

    ctx.feedback("ambiguous")
        .query(ambiguous)
        .rank(|_| FeedbackRank::Names)
        .display(|_db, writer, _, (_, name, definitions)| {
            writer.code(name.to_string());
            writer.string(" could refer to ");
            writer.list("or", |list| {
                for &definition in *definitions {
                    list.add(move |writer| writer.node(definition));
                }
            });
            writer.string(".");
            writer.line_break();
            writer.string("Rename the extra definitions.");
        })
        .register();
}
